pkg/services/grafana-apiserver/rest/dualwriter.go

```diff
@@ -79,35 +89,92 @@ func (d *DualWriter) Create(ctx context.Context, obj runtime.Object, createValid
 // Update overrides the default behavior of the Storage and writes to both the LegacyStorage and Storage.
 func (d *DualWriter) Update(ctx context.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc, forceAllowCreate bool, options *metav1.UpdateOptions) (runtime.Object, bool, error) {
 	if legacy, ok := d.legacy.(rest.Updater); ok {
-		_, _, err := legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+		// Will resource version checking work????
+		old, err := d.Get(ctx, name, &metav1.GetOptions{})
+		if err != nil {
+			return nil, false, err
+		}
+		accessor, err := meta.Accessor(old)
+		if err != nil {
+			return nil, false, err
+		}
+		// Hold on to the RV+UID for the dual write
+		theRV := accessor.GetResourceVersion()
+		theUID := accessor.GetUID()
+
+		// Changes applied within new storage
+		// will fail if RV is out of sync
+		updated, err := objInfo.UpdatedObject(ctx, old)
+		if err != nil {
+			return nil, false, err
+		}
+
+		accessor, err = meta.Accessor(updated)
+		if err != nil {
+			return nil, false, err
+		}
+		accessor.SetUID("")             // clear it
+		accessor.SetResourceVersion("") // remove it so it is not a constraint
+		obj, created, err := legacy.Update(ctx, name, &updateWrapper{
+			upstream: objInfo,
+			updated:  updated, // returned as the object that will be updated
+		}, createValidation, updateValidation, forceAllowCreate, options)
+		if err != nil {
+			return obj, created, err
+		}
+
+		accessor, err = meta.Accessor(obj)
 		if err != nil {
 			return nil, false, err
 		}
+		accessor.SetResourceVersion(theRV) // the original RV
+		accessor.SetUID(theUID)
+		objInfo = &updateWrapper{
+			upstream: objInfo,
+			updated:  obj, // returned as the object that will be updated
+		}
 	}
 
 	return d.Storage.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
 }
 
 // Delete overrides the default behavior of the Storage and delete from both the LegacyStorage and Storage.
 func (d *DualWriter) Delete(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
-	if legacy, ok := d.legacy.(rest.GracefulDeleter); ok {
-		_, _, err := legacy.Delete(ctx, name, deleteValidation, options)
-		if err != nil {
-			return nil, false, err
+	// Delete from storage *first* so the item is still exists if a failure happens
```

Comment:
not tested in this PR, but I think we should flip the direction of possible errors.  keeping the value in SQL as long as possible